### PR TITLE
Improve the user course enrollments form - give better reasons why so…

### PR DESCRIPTION
**Pull request by mehalshah**
…me assignment fails and enroll everything you can

### CODE REVIEW

File: dashboard/app/controllers/plc/user_course_enrollments_controller.rb [hunk 1]
```diff
@@ -34,26 +34,23 @@ def new
...
+    notice_string += "The following users were not teachers #{listify(nonteacher_users)}<br/>" unless nonteacher_users.empty?
+    notice_string += "The following users failed for other reasons #{listify(other_failure_users)}<br/>" unless other_failure_users.empty?
 
-    redirect_to action: :new, notice: "Enrollments created for #{user_emails.join(', ')}"
+    redirect_to action: :new, notice: notice_string
   end
 
   private
 
+  def listify(user_list)
```
Comment (aoby) [inline dashboard/app/controllers/plc/user_course_enrollments_controller.rb:52]: nice

File: dashboard/app/controllers/plc/user_course_enrollments_controller.rb [hunk 2]
```diff
@@ -34,26 +34,23 @@ def new
...
+    user_list.map {|user| "<li>#{user}</li>"}.join ''
```
Comment (aoby) [inline dashboard/app/controllers/plc/user_course_enrollments_controller.rb:53]: Will this work in the above html without a surrounding `<ul>` or similar?

Comment (mehalshah) [inline dashboard/app/controllers/plc/user_course_enrollments_controller.rb:53]: Yup

File: dashboard/app/models/plc/user_course_enrollment.rb [hunk 3]
```diff
@@ -28,6 +28,33 @@ class Plc::UserCourseEnrollment < ActiveRecord::Base
...
+      if user.nil?
+        nonexistant_users << email
+      elsif !user.teacher?
+        nonteacher_users << email
+      else
+        enrollment = find_or_create_by(user: user, plc_course: course)
+        if enrollment.valid?
+          enrolled_users << email
+        else
+          other_failure_users << email
```
Comment (aoby) [inline dashboard/app/models/plc/user_course_enrollment.rb:51]: In this other case, do we have more details as to why it's invalid? Should we display the `enrollment.errors` to user?

File: dashboard/test/models/plc/user_course_enrollment_test.rb [hunk 4]
```diff
@@ -14,4 +14,19 @@ class Plc::UserCourseEnrollmentTest < ActiveSupport::TestCase
...
+    @student.update(email: student_email)
+    nonexistant_email = 'wrong-email@wrong.com'
+
+    created_enrollments, nonexistant_users, nonteacher_users, other_failure_users =
+      Plc::UserCourseEnrollment.enroll_users([@user.email, nonexistant_email, student_email], @course.id)
+
+    assert_equal created_enrollments, [@user.email]
+    assert_equal nonexistant_users, [nonexistant_email]
+    assert_equal nonteacher_users, [student_email]
+    assert_empty other_failure_users
```
Comment (aoby) [inline dashboard/test/models/plc/user_course_enrollment_test.rb:30]: Can we have a test for the other failures too? Even if you mock `UserCourseEnrollment.valid?` to return false.

**Comment by aoby** [comment]
Few small suggestions, otherwise LGTM

File: dashboard/app/models/plc/user_course_enrollment.rb (hunk 3 above)
Comment (mehalshah) [inline dashboard/app/models/plc/user_course_enrollment.rb:51]: Will do

**Pull request merged**